### PR TITLE
ET-1936 main work

### PR DESCRIPTION
**Pull request by lucatume**
Based on [Common PR](https://github.com/the-events-calendar/tribe-common/pull/2050)

Ticket: [ET-1936](https://stellarwp.atlassian.net/browse/ET-1936)

[Demo](https://share.cleanshot.com/t9bKkmTZ)

This PR lays the backend base to support multiple check-ins of Series Pass Attendees.
The objective is to allow a Series Pass Attendee to check-in to each Event part of the Series with the same QR code.
The site administrator would, then have access to per-Event Attendees and Attendance details.

This is a pretty big feature, I'm leaving breadcrumbs:
1. Starting from the end, Attendees will be pulled on the Event, actually Occurrence, Attendees screen with a repository query similar to `tribe_attendees()->where( 'event', $occurrence_provisional_id )->all()`.
2. On the Occurrence Attendee list should show: the Series Pass Attendees that have not yet checked-in and the checked-in Occurrence Attendees (clones, more on this later).
3. The Attendees Controller intercepts the "filter-by-event" query above to pull Attendees for the Series and the Occurrence, [taking care not to fetch Attendees that have been cloned and related with the Occurrence.](https://github.com/the-events-calendar/event-tickets/blob/bd5a732996f748b37a0d343330aa199ec1869d14/src/Tickets/Flexible_Tickets/Series_Passes/Attendees.php#L1102-L1120)
4. How did those cloned Attendee came to be? When a Series Pass Attendee checks into an Occurrence part of a Series it's not actually the Series Pass Attendee that is checked in, or updated, [but a clone created on demand](https://github.com/the-events-calendar/event-tickets/blob/bd5a732996f748b37a0d343330aa199ec1869d14/src/Tickets/Flexible_Tickets/Series_Passes/Attendees.php#L253) that is [related to the original by means of the meta key used in the query above.](https://github.com/the-events-calendar/event-tickets/blob/bd5a732996f748b37a0d343330aa199ec1869d14/src/Tickets/Flexible_Tickets/Series_Passes/Attendees.php#L439-L440).
5. Note the code is doing this cloning both when the check-in happens using a QR code, or when the check-in is done manually, from the Admin UI.
6. The difference is that, when the Attendee is checked in from the QR, [the check-in restrictions will apply, if present.](https://github.com/the-events-calendar/event-tickets/blob/bd5a732996f748b37a0d343330aa199ec1869d14/src/Tickets/Flexible_Tickets/Series_Passes/Attendees.php#L589-L595)
7. Since the cloned Attendees will be related to Occurrence Provisional IDs bound to be updated when the base changes, the PR adds a functionality to the Base Controller to make sure Attendees related to Occurrences by mean of their provisional IDs [will be updated in the background.](https://github.com/the-events-calendar/event-tickets/blob/b380838c7ede36845c510cbe34235f938cc0a2ca/src/Tickets/Flexible_Tickets/Base.php#L236)
8. The above functionality is only used, right now, by Attendees created as clones of Series Pass Attendees, but will be later required for Standard Tickets on Occurrences, I've just already moved the functionality where it belongs. It's [tested in the Attendees Controller as it's the only one using it.](https://github.com/the-events-calendar/event-tickets/blob/bd5a732996f748b37a0d343330aa199ec1869d14/tests/ft_integration/TEC/Tickets/Flexible_Tickets/Series_Passes/AttendeesTest.php#L1187)
9. When a Series Pass Attendee is cloned during manual check-in from the UI, then this will trigger [a reload of the Attendees list](https://github.com/the-events-calendar/event-tickets/blob/ad3b1e06df1c0d1d07c84021ee2b0a6109f6e216/src/resources/js/tickets-attendees.js#L145-L147) thanks to a [per-request (i.e. the AJAX request) tracking](https://github.com/the-events-calendar/event-tickets/blob/ad3b1e06df1c0d1d07c84021ee2b0a6109f6e216/src/Tickets/Flexible_Tickets/Series_Passes/Attendees.php#L1203-L1209).

A note about the choice to relate cloned (Occurrence) Attendees to the original Series Attendee with a meta key: the `post_parent` field is not available across all Attendees from all ticket providers; e.g. Tickets Commerce uses the `post_parent` field of the Attendee to relate it to the Order.
Given this limit, and to avoid having to create a custom table, the code walks back to lay the ground for an efficient query (all indexed, point 3.

A second note: the code lives in ET, but will work for Attendees from all ticket providers [by leveraging the methods on the `tribe_attendees` object](https://github.com/the-events-calendar/event-tickets/blob/bd5a732996f748b37a0d343330aa199ec1869d14/src/Tickets/Flexible_Tickets/Series_Passes/Attendees.php#L1092) to pull all the current Attendee post types, meta keys and so on.

[ET-1936]: https://stellarwp.atlassian.net/browse/ET-1936?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: tests/ft_integration/Tribe/Tickets/AttendeesTest.php [hunk 1]
```diff
@@ -58,66 +59,66 @@ private function get_attendee_data( array $attendees ): array {
...
+//				$event_id = tribe_events()->set_args( [
+//					'title'      => 'Event with no attendees',
+//					'status'     => 'publish',
+//					'start_date' => '2020-01-01 00:00:00',
+//					'duration'   => 2 * HOUR_IN_SECONDS,
+//				] )->create()->ID;
+//				$ticket_id = $this->create_rsvp_ticket( $event_id );
+//				$this->create_many_attendees_for_ticket( 3, $ticket_id, $event_id );
+//
+//				return [ $event_id, [ $event_id, $ticket_id ] ];
```
Comment (rafsuntaskin) [inline tests/ft_integration/Tribe/Tickets/AttendeesTest.php:102]: We are removing these?

Comment (lucatume) [reply tests/ft_integration/Tribe/Tickets/AttendeesTest.php:102]: Most definitely not, my bad.

File: src/Tribe/Metabox.php [hunk 2]
```diff
@@ -722,11 +713,22 @@ public function ajax_attendee_checkin() {
...
 
-		wp_send_json_success( $did_checkin );
+		$data = [ 'did_checkin' => $did_checkin ];
+		/**
+		 * Filters the status code to return when an attendee is checked in.
+		 *
+		 * @since TBD
+		 *
+		 * @param int $status      The HTTP status code to return. Default is 200.
+		 * @param int $attendee_id The ID of the attendee that was checked in.
```
Comment (rafsuntaskin) [inline src/Tribe/Metabox.php:727]:
```suggestion
		 * @param array{did_checkin: bool} $data        The AJAX response data.
		 * @param int                      $attendee_id The ID of the attendee that was checked in.
```

File: src/Tribe/REST/V1/Endpoints/QR.php [hunk 3]
```diff
@@ -245,16 +247,16 @@ public function check_in( WP_REST_Request $request ) {
 		}
 
 		$event_id      = (int) $qr_arr['event_id'];
-		$ticket_id     = (int) $qr_arr['ticket_id'];
+		$attendee_id     = (int) $qr_arr['ticket_id'];
```
Comment (rafsuntaskin) [inline src/Tribe/REST/V1/Endpoints/QR.php:250]:
```suggestion
		$attendee_id    = (int) $qr_arr['ticket_id'];
```

**Review by rafsuntaskin — COMMENTED** [review]
Minor updates otherwise looking good.

**Review by rafsuntaskin — APPROVED** [review]
Minor updates otherwise looking good.

**Pull request merged**